### PR TITLE
[C API] Add APIs for getting/setting function types, and a CallRef example

### LINKED ISSUE #6718: How to BinaryenCallRef
Hello I am trying to use BinaryenCallRef but its not clear how. Unfortunately there is no example in c-api-kitchen-sink or other docs.
I tried it this way:
```cpp
  BinaryenAddFunction(mod, "name", BinaryenTypeCreate(nullptr, 0), BinaryenTypeCreate(nullptr, 0), nullptr, 0, BinaryenUnreachable(mod));
  auto exp = BinaryenRefFunc(mod, "name", BinaryenTypeFuncref());
  exp      = BinaryenCallRef(mod, exp, nullptr, 0, BinaryenTypeFuncref(), false);

  BinaryenAddFunction(mod, "name2", BinaryenTypeCreate(nullptr, 0), BinaryenTypeCreate(nullptr, 0), nullptr, 0, exp);
   
```
But the end boss is always [this](https://github.com/WebAssembly/binaryen/blob/main/src/wasm/wasm.cpp#L995) assert,

Any tips how I should use it or should I go back to `call_indirect` for now?

### PR DESCRIPTION
**Pull request by kripken**
Fixes #6718

### CODE REVIEW

File: test/example/c-api-kitchen-sink.c
Comment (tlively) [inline test/example/c-api-kitchen-sink.c:1]: What would you think about putting this test in a gtest file? Compilation errors in the gtests are caught during the normal build and the tests run much faster, so they seem strictly better than the example tests.

**Review by tlively — APPROVED** [review]

File: test/example/c-api-kitchen-sink.c
Comment (kripken) [reply test/example/c-api-kitchen-sink.c:1]: I agree this file might make sense as a gtest, yeah. There was the idea of having special "example" tests that are helpful for people to see code samples of the public API (as opposed to testing of internals, or corner cases of the public API), which is what "example" was meant to invoke, but I suppose we could make `gtest/example` and document that?

Comment (tlively) [reply test/example/c-api-kitchen-sink.c:1]: Yeah, gtest/example would be great.

Comment (kripken) [reply test/example/c-api-kitchen-sink.c:1]: I started to do this but then realized a big missing feature of our gtests atm, which is that they don't work with the auto updater script. So tests like this that match the text format output would need to be manually updated if we ever change the text format. Before we move any significant amount of such tests to gtest we should probably add auto-updating.

Comment (kripken) [reply test/example/c-api-kitchen-sink.c:1]: (Which I agree is worth doing, but not trivial.)

**Pull request merged**